src/Agent.Worker/TestResults/ResultsCommandExtension.cs

```diff
@@ -518,6 +195,155 @@ private void AddTargetBranchInfoToRunCreateModel(RunCreateModel runCreateModel,
+                _platform = _configuration = null;
+            }
+
+            if (!string.IsNullOrWhiteSpace(_executionContext.Variables.Release_ReleaseUri))
+            {
+                releaseUri = _executionContext.Variables.Release_ReleaseUri;
+                releaseEnvironmentUri = _executionContext.Variables.Release_ReleaseEnvironmentUri;
+            }
+
+            // If runName is not provided by the task, then create runName from testRunner name and buildId.
+            string runName = String.IsNullOrWhiteSpace(_runTitle)
+                ? String.Format("{0}_TestResults_{1}", _testRunner, buildId)
+                : _runTitle;
+
+            StageReference stageReference = new StageReference() { StageName = stageName, Attempt = Convert.ToInt32(stageAttempt) };
+            PhaseReference phaseReference = new PhaseReference() { PhaseName = phaseName, Attempt = Convert.ToInt32(phaseAttempt) };
+            JobReference jobReference = new JobReference() { JobName = jobName, Attempt = Convert.ToInt32(jobAttempt) };
+            PipelineReference pipelineReference = new PipelineReference()
+            {
+                PipelineId = buildId,
+                StageReference = stageReference,
+                PhaseReference = phaseReference,
+                JobReference = jobReference
+            };
+
+            TestRunContext testRunContext = new TestRunContext(
+                owner: owner,
+                platform: _platform,
+                configuration: _configuration,
+                buildId: buildId,
+                buildUri: buildUri,
+                releaseUri: releaseUri,
+                releaseEnvironmentUri: releaseEnvironmentUri,
+                runName: runName,
+                testRunSystem: _testRunSystem,
+                buildAttachmentProcessor: new CodeCoverageBuildAttachmentProcessor(),
+                targetBranchName: pullRequestTargetBranchName,
+                pipelineReference: pipelineReference
+            );
+            return testRunContext;
+
+        }
+
+        private PublishOptions GetPublishOptions()
+        {
+            var publishOptions = new PublishOptions()
+            {
+                IsMergeTestResultsToSingleRun = _mergeResults,
+                IsAddTestRunAttachments = _publishRunLevelAttachments
+            };
+
+            return publishOptions;
+        }
+
+        private async Task PublishTestRunDataAsync(VssConnection connection, String teamProject, TestRunContext testRunContext)
+        {
+            bool isTestRunOutcomeFailed = false;
+            try
+            {
+                var featureFlagService = HostContext.GetService<IFeatureFlagService>();
+                featureFlagService.InitializeFeatureService(_executionContext, connection);
+
+                if (featureFlagService.GetFeatureFlagState(_publishTestResultsLibFeatureFlag, Service.TFS)){
+                    TestDataProvider testDataProvider = ParseTestResultsFile(testRunContext);
```

Comment:
>TestDataProvider testDataProvider = ParseTestResultsFile(testRunContext); [](start = 20, length = 73)

Can we move this also inside TestRunDataPublisher, and we can accept the files list as input, like older APIs. So there will be very less code in this file.